### ISSUE TITLE
PackedCoordinateSequenceFactory does not work correctly when dealing with CoordinateXY.

### ISSUE THREAD

**Issue by pascalriquier** [locationtech/jts#434]
PackedCoordinateSequence$Double calls getOrdinate(2) or getOrdinate(3) on the coordinates and that throws an exception in CoordinateXY. 

`		Polygon polygonA = new GeometryFactory(new PackedCoordinateSequenceFactory()).createPolygon(
			new Coordinate[] {
				new CoordinateXY(0, 0),
				new CoordinateXY(10, 0),
				new CoordinateXY(10, 10),
				new CoordinateXY(0, 10),
				new CoordinateXY(0, 0)
			}
		);

		Polygon polygonB = new GeometryFactory().createPolygon(
				new Coordinate[] {
						new CoordinateXY(5, 5),
						new CoordinateXY(15, 5),
						new CoordinateXY(15, 15),
						new CoordinateXY(5, 15),
						new CoordinateXY(5, 5)
				}
		);
		
		polygonB.intersection(polygonA);
		polygonA.intersection(polygonB);
`
In the above example, the first intersection does not throw an exception while the second does. 
PackedCoordinateSequence$Float does this differently by calling getZ() on the coordinates. So possibly the solution is to call getZ() from getOrdinate() in CoordinateXY.

**Comment by dr-jts** [comment]
The root cause of this is that internally the intersection algorithm winds up working with an array of Coordinates of mixed dimension.  Also, there isn't enough metadata around to determine that the output should be XY, not XYZ.

Not sure how to fix this.  I suspect it would take quite a bit to figure out how to get everything to work correctly with the right dimensions and coordinate types where needed.

Part of the problem is the decision to throw error when accessing dimensions that do not exist in CoordinateXY (and other Coordinate subclasses).  I know @jodygarnett has the opinion that it's better to fail fast, but if we don't understand how to prevent this problem, then it may be better to Do No Harm and simply return NaN.

**Comment by dr-jts** [comment]
> . So possibly the solution is to call getZ() from getOrdinate() in CoordinateXY.

I don't think this will work, because the `getOrdinate(2)` call may be accessing either Z or M depending on the source.

**Comment by dr-jts** [comment]
Further to the idea of making invalid ordinate access succeed, this is already done in `CoordinateXY.getZ()`.  So making `getOrdinate(2)` return NaN would be consistent.
